### PR TITLE
Add support for non-seekable streaming uploads

### PR DESCRIPTION
**Pull request by JordonPhillips**
The tricky part of this change is that existing code relies on the
ability to get the length of the object being uploaded, which is
not possible for non-seekable objects. To work around that issue,
this will read data into memory up to the configured multipart
threshold to determine if it requires a multipart upload. That data
will then be used to generate the initial parts.

cc @kyleknap @jamesls

### CODE REVIEW

File: tests/functional/test_upload.py [hunk 1]
```diff
@@ -296,6 +297,17 @@ def test_upload_for_seekable_filelike_obj(self):
         self.assert_expected_client_calls_were_correct()
         self.assert_upload_part_bodies_were_correct()
 
+    def test_upload_for_non_seekable_filelike_obj(self):
```
Comment (kyleknap) [inline tests/functional/test_upload.py]: We probably will want to add one for the non mulitpart upload case as well.

File: s3transfer/upload.py [hunk 2]
```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
...
+class UploadNonSeekableInputManager(UploadInputManager):
+    """Upload utility for a file-like object that cannot seek."""
+    def __init__(self, osutil, transfer_coordinator):
+        super(UploadNonSeekableInputManager, self).__init__(
+            osutil, transfer_coordinator)
+        self._initial_data = None
+
+    @classmethod
+    def is_compatible(cls, upload_source):
+        return not seekable(upload_source)
```
Comment (kyleknap) [inline s3transfer/upload.py]: So I am not sure if the qualifier here should be weather it is seekable or not. Technically, you could put a seekable file through this Manager and it should work. I think the compatibility here should be if the file is readable as if it fails that the ``UploadSubmissionTask`` will throw an error as this is the last one in the input manager resolver chain.

File: s3transfer/upload.py [hunk 3]
```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
...
+        # No-op because there is no way to do this short of reading the entire
+        # body into memory.
+        return
+
+    def requires_multipart_upload(self, transfer_future, config):
+        # This is tricky to determine in this case because we can't know how
+        # large the input is. So to figure it out, we read data into memory
+        # up until the threshold and compare how much data was actually read
+        # against the threshold.
+        fileobj = transfer_future.meta.call_args.fileobj
```
Comment (kyleknap) [inline s3transfer/upload.py]: If the user provided a size themselves, I would say we do not stream the file in. So I would make sure you have this check at beginning: https://github.com/boto/s3transfer/blob/develop/s3transfer/upload.py#L186 and then do the streaming.

File: s3transfer/upload.py [hunk 4]
```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
...
+        callbacks = get_callbacks(transfer_future, 'progress')
+        part_number = 0
+
+        # First go over the parts generated by calling
+        # `requires_multipart_upload` if any.
+        if self._initial_data is not None:
+            parts = self._yield_first_parts(transfer_future, config, callbacks)
+            for part in parts:
+                part_number += 1
+                yield part_number, part
```
Comment (kyleknap) [inline s3transfer/upload.py]: We talked about this but it would be great if we had all of this logic under one loop and one yield from. The two seperate loops and the three places to yield from is a bit complicated.

File: s3transfer/upload.py [hunk 5]
```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
...
+        :type config: s3transfer.manager.TransferConfig
+        :param config: The config associated to the transfer manager.
+
+        :type callbacks: list
+        :param callbacks: The callbacks associated with the transfer future.
+
+        :return: Generator which generates part bodies from the initial data.
+        """
+        try:
+            fileobj = transfer_future.meta.call_args.fileobj
```
Comment (kyleknap) [inline s3transfer/upload.py]: These three lines do not need to be in the try except block.

File: s3transfer/upload.py [hunk 6]
```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
...
+                    yield self._wrap_part_data(
+                        self._initial_data[i:part_bound], callbacks)
+        except GeneratorExit as e:
+            # Zero out the initial data now that we're done generating parts
+            # from it so that we don't hold onto those bytes any longer
+            # than we need to.
+            self._initial_data = None
+            raise e
+
+    def _wrap_part_data(self, data, callbacks, size=None):
```
Comment (kyleknap) [inline s3transfer/upload.py]: I think we can call it something more general like ``_wrap_data`` as this is used for PutObject and UploadPart.

File: tests/integration/test_upload.py [hunk 7]
```diff
@@ -26,22 +27,23 @@ def setUp(self):
         self.config = TransferConfig(
             multipart_threshold=self.multipart_threshold)
 
+    def create_file(self, size, name=''):
```
Comment (kyleknap) [inline tests/integration/test_upload.py]: Let's call it ``get_input_fileobj()`` or something along those lines. ``create_file`` for me seems like you are intending to create a file each time and in reality for the file-like objects you are just creating a file-like object without writing to disk.

File: tests/unit/test_upload.py [hunk 8]
```diff
@@ -262,6 +266,71 @@ def test_stores_bodies_in_memory_upload_part(self):
...
+
+        parts = list(self.upload_input_manager.yield_upload_part_bodies(
+            self.future, self.config))
+
+        size = self.config.multipart_chunksize
+        num_parts = len(self.content) / size
+        num_parts = math.ceil(num_parts)
+        self.assertEqual(len(parts), num_parts)
+
+        for i, part in enumerate(parts[:-1]):
```
Comment (kyleknap) [inline tests/unit/test_upload.py]: I can follow this but it took a little bit of time. Comments would have helped especially in the for loop, the initial part calculations, and the final part assertion.

**Comment by kyleknap** [comment]
Implementation looks good. There was a little bit of refactoring to simplify the generator that I pointed out that we can do and a couple edge cases that looks like you may have missed, but looks good in general.

**Comment by JordonPhillips** [comment]
Updated

**Comment by kyleknap** [comment]
Looks like python3 tests are failing...

**Comment by JordonPhillips** [comment]
Fixed

**Comment by kyleknap** [comment]
Nice. Really like the refactoring you did. Makes it much easier to follow. :ship:

**Comment by kyleknap** [comment]
Also before merging, it would not hurt to do a performance test to see how the speed compares to the seekable and filename case.

**Comment by JordonPhillips** [comment]
Rebased and squashed commits

**Comment by JordonPhillips** [comment]
Performance differences seem to be negligible.

Time in seconds for a single 7 byte upload:

| Actual File                   | Seekable File-Like         | Non-Seekable File-Like |
| ----------------------------- | ------------------------------- | ------------------------------- |
| 0.1925144976199954 | 0.22886600455996814 | 0.15985330105002504 |

Time in seconds for a multipart upload of 5MB + 1 byte with a 5MB chunksize and threshold:

| Actual File                   | Seekable File-Like         | Non-Seekable File-Like |
| ----------------------------- | ------------------------------- | ------------------------------- |
| 0.8002007092098938 | 0.7937404662198969 | 0.8091048456198768 |

**Comment by kyleknap** [comment]
Thanks. :ship:
